File: src/my_cpp_pkg/src/number_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/int64.hpp"

class NumberPublisher : public rclcpp::Node
{
public:
    /** ROS 2 基础 C++ 节点构造函数 */
    NumberPublisher() : Node("number_publisher"), number_(2)
    {
        publisher_ = this->create_publisher<std_msgs::msg::Int64>("number", 10);
        // 创建定时器，每 1 秒发布一次数字
        number_timer_ = this->create_timer(
            std::chrono::seconds(1),
            std::bind(&NumberPublisher::publish_number, this)
        );
        RCLCPP_INFO(this->get_logger(), "Number publisher node has been started.");
    }

private:
    /** 发布者 */
    rclcpp::Publisher<std_msgs::msg::Int64>::SharedPtr publisher_;
    /** 要发布的数字 */
    int64_t number_;
    /** 定时器 */
    rclcpp::TimerBase::SharedPtr number_timer_;
    /** 发布数字 */
    void publish_number()
    {
        auto message = std_msgs::msg::Int64();
        message.data = number_;
        publisher_->publish(message);
        number_++;
        RCLCPP_INFO(this->get_logger(), "Publishing number: %ld", message.data);
    }
};

int main(int argc, char** argv)
{
    // 初始化 rclcpp
    rclcpp::init(argc, argv);
    // 创建节点智能指针实例（ROS 2 推荐使用智能指针）
    auto node = std::make_shared<NumberPublisher>();
    // 保持节点运行（阻塞式）
    rclcpp::spin(node);
    // 关闭 rclcpp，释放资源
    rclcpp::shutdown();
    return 0;
}

File: src/my_py_pkg/my_py_pkg/number_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64  # 导入整数消息类型

class NumberPublisherNode(Node):
    """ROS 2 基础 Python 节点类"""
    def __init__(self):
        # 初始化父类，设置节点名称
        super().__init__("number_publisher")
        # 可在此处添加自定义初始化逻辑（如定时器、发布者、订阅者等）
        self.number_ = 2
        self.number_publisher_ = self.create_publisher(Int64, "number", 10)
        self.number_timer_ = self.create_timer(1.0, self.publish_number)
        self.get_logger().info("Number publisher has been started.")

    def publish_number(self):
        msg = Int64()
        msg.data = self.number_
        self.number_publisher_.publish(msg)
        self.get_logger().info(f"Published number: {self.number_}")

def main(args=None):
    """节点主函数"""
    # 初始化 rclpy
    rclpy.init(args=args)
    # 创建节点实例
    node = NumberPublisherNode()
    # 保持节点运行（阻塞式）
    rclpy.spin(node)
    # 关闭 rclpy，释放资源
    rclpy.shutdown()

if __name__ == "__main__":
    # 脚本直接运行时执行主函数
    main()